IECoreAlembic : Attribute support

donboie opened the pull request:
**Types which are converted on load:**

- V2s -> V2i
- V3s -> V3i
- P2s -> V2i ( point )
- P3s -> V3i ( point )
- Box2s -> Box2i
- Box3s -> Box3i
- C3c -> Color3f ( 0 -> 255 remapped to 0 -> 1 )
- C4c -> Color4f ( 0 -> 255 remapped to 0 -> 1 )

**Unsupported types**

- Wstring

johnhaddon commented on contrib/IECoreAlembic/src/IECoreAlembic/AlembicScene.cpp:
```diff
@@ -267,6 +319,493 @@ class AlembicScene::AlembicReader : public AlembicIO
+
+			auto timeSampling = propertyHeader->getTimeSampling();
+
+			return sampleInterval( timeSampling.get(), timeSampling->getNumStoredTimes(), time, floorIndex, ceilIndex );
```
Again here, the number of samples needs to come frm the property, not `timeSampling->getNumStoredTimes()`.

johnhaddon commented on contrib/IECoreAlembic/src/IECoreAlembic/AlembicScene.cpp:
```diff
@@ -267,6 +319,493 @@ class AlembicScene::AlembicReader : public AlembicIO
+
+			CompoundPropertyReaderPtr propertyReader = GetCompoundPropertyReaderPtr( userProperties );
+
+			ScalarPropertyReaderPtr scalarPropertyReader = propertyReader->getScalarProperty( name.string() );
```
Ooh, I didn't know about this. So if I understand correctly, this is a lower-level less typesafe thing that avoids us having to "downcast" to various different `ITypedScalarProperty<Something>` properties? It looks like it also has  `isConstant()` and `getNumSamples()` methods, so that could help out with the comments I've made above.

johnhaddon commented on contrib/IECoreAlembic/src/IECoreAlembic/AlembicScene.cpp:
```diff
@@ -267,6 +319,493 @@ class AlembicScene::AlembicReader : public AlembicIO
+			CompoundPropertyReaderPtr propertyReader = GetCompoundPropertyReaderPtr( userProperties );
+
+			ScalarPropertyReaderPtr scalarPropertyReader = propertyReader->getScalarProperty( name.string() );
+			DataType dataType = scalarPropertyReader->getDataType();
```
Assigning these to a variable forces a reader to wonder if they are going to be modified, and then look through the rest of this 400 line function to check that they're not. Is this an attempt at improving clarity or is it important for performance? If for clarity, I would suggest that the variable isn't really helping, at least for me. If it's for performance (or if you strongly prefer them anyway), could you make them const so it's obvious they won't change?

donboie replied:
>  Is this an attempt at improving clarity or is it important for performance?

Attempt to reduce line length. I started with reading those values when required but refactored to this unclear state. Perhaps I should reduce the size of this function, it's a bit of a beast.

johnhaddon commented on contrib/IECoreAlembic/src/IECoreAlembic/AlembicScene.cpp:
```diff
@@ -267,6 +319,493 @@ class AlembicScene::AlembicReader : public AlembicIO
+			PlainOldDataType pod = dataType.getPod();
+			uint8_t extent = dataType.getExtent();
+			const MetaData &metaData = scalarPropertyReader->getMetaData();
+			std::string interpretation = metaData.get( "interpretation" );
```
Actually, it looks like here we're allocating a string that isn't even used in many of the code paths? To me it seems it would be better just to get the interpretation directly where it is needed.

johnhaddon commented on contrib/IECoreAlembic/src/IECoreAlembic/AlembicScene.cpp:
```diff
@@ -267,6 +319,493 @@ class AlembicScene::AlembicReader : public AlembicIO
+			}
+			else if( pod == kUint16POD )
+			{
+				unsigned short value;
```
Should we be using `uint16_t`, `uint8_t` here instead of `char`, `short` etc?

johnhaddon commented on contrib/IECoreAlembic/src/IECoreAlembic/AlembicScene.cpp:
```diff
@@ -267,6 +319,493 @@ class AlembicScene::AlembicReader : public AlembicIO
+			}
+			else if( pod == kUint64POD )
+			{
+				unsigned int value;
```
`uint64_t`?

johnhaddon commented on contrib/IECoreAlembic/src/IECoreAlembic/AlembicScene.cpp:
```diff
@@ -267,6 +319,493 @@ class AlembicScene::AlembicReader : public AlembicIO
+			}
+			else if( pod == kInt64POD )
+			{
+				int tmpInt;
```
`int64_t`? I'll let you check the others...

johnhaddon commented on contrib/IECoreAlembic/src/IECoreAlembic/AlembicScene.cpp:
```diff
@@ -267,6 +319,493 @@ class AlembicScene::AlembicReader : public AlembicIO
+				return;
+			}
+
+			h.append( fileName() );
```
Can we avoid hashing these in when there are no user properties? It seems that could be quite a common case, and it'd be well worth it in terms of downstream performance.

johnhaddon commented on contrib/IECoreAlembic/src/IECoreAlembic/AlembicScene.cpp:
```diff
@@ -875,6 +1424,262 @@ class AlembicScene::AlembicWriter : public AlembicIO
+				);
+			}
+
+			if( !runTimeCast<const IECore::Data>( attribute ) )
```
Isn't this redundant given the block at line 1670?

johnhaddon commented on contrib/IECoreAlembic/src/IECoreAlembic/AlembicScene.cpp:
```diff
@@ -875,6 +1424,262 @@ class AlembicScene::AlembicWriter : public AlembicIO
+			}
+			else
+			{
+				throw IECore::Exception(
```
Throwing here will prevent us from using the SceneWriter with any scene that has shader assignments. I suggest we just make it a warning message instead...

donboie commented:
> Do you have any performance numbers for this? For instance, how much does it affect reading of a scene which doesn't have attributes at all? And what's the reading overhead of adding a bunch of attributes? 

I'll generate some numbers to understand the performance impact of attribute reads. Gives me a chance to use my fancy parallel read testing function!

> What I found when writing this stuff in the first place was that it was really expensive to construct the various schemas and suchlike, and we were much much better off getting them once and storing them in the AlembicIO classes. I'm wondering if the same might apply here, or if the properties you're dealing with have lower overhead.

Do you have any numbers from your initial investigation?

donboie commented:
> Do you have any performance numbers for this? For instance, how much does it affect reading of a scene which doesn't have attributes at all?

There is no impact on scene reading when no attributes are present.

>  And what's the reading overhead of adding a bunch of attributes?

d16dc74 & 3507af9 add some tests to verify the performance of attribute reads. 100,000 leaf locations with 4 attributes (float, int ,V3f and small string)

test case             | parallel read time /s 
-----------------------|-------
.abc no attributes | 1.07 
.abc attributes      | 3.74
.scc no attributes | 0.53
.abc attributes      |4.65 

> Note: Alembic compared against the .scc attribute read case also to give a little more context to the numbers

Also tried the typed property readers and noticed that there performance didn't change. I'll use these rather than the lower level, untyped *ScalarPropertyReader* as it sould be easier to refactor the read case to use templates and reduce the LOC in the readAttribute function.

run | typed parallel read time / s | untyped parallel read time /s 
-----| ------------------------------------|---------------------------------------
0    | 3.74                                    | 3.63
1    | 3.9                                      | 3.66
2    | 3.82                                    | 4.08
3    | 3.7                                      | 3.66
4    | 3.86                                    | 4.04
mean | 3.804                              | 3.814

donboie commented:
> Also tried the typed property readers and noticed that there performance didn't change. I'll use these rather than the lower level, untyped ScalarPropertyReader as it sould be easier to refactor the read case to use templates and reduce the LOC in the readAttribute function.

I found this didn't really help as much as I'd thought it would.

Hopefully I've addressed your comments @johnhaddon.

andrewkaufman commented on contrib/IECoreAlembic/src/IECoreAlembic/AlembicScene.cpp:
```diff
@@ -267,6 +312,601 @@ class AlembicScene::AlembicReader : public AlembicIO
+
+			for( size_t i = 0; i < userProperties.getNumProperties(); ++i )
+			{
+				attrs.push_back( InternedString( userProperties.getPropertyHeader( i ).getName() ) );
```
Should we be prefixing with “user:” ?

andrewkaufman commented on contrib/IECoreAlembic/src/IECoreAlembic/AlembicScene.cpp:
```diff
@@ -267,6 +312,601 @@ class AlembicScene::AlembicReader : public AlembicIO
+				return false;
+			}
+
+			const AbcA::PropertyHeader *propertyHeader = userProperties.getPropertyHeader( name.string() );
```
Should we be stripping any prefix? Or just the "user:" prefix?

andrewkaufman commented on contrib/IECoreAlembic/src/IECoreAlembic/AlembicScene.cpp:
```diff
@@ -267,6 +312,601 @@ class AlembicScene::AlembicReader : public AlembicIO
+				return 0;
+			}
+
+			ScalarPropertyReaderPtr scalarPropertyReader = propertyReader->getScalarProperty( name.string() );
```
Should we be stripping any prefix? Or just the "user:" prefix?

andrewkaufman commented on contrib/IECoreAlembic/src/IECoreAlembic/AlembicScene.cpp:
```diff
@@ -267,6 +312,601 @@ class AlembicScene::AlembicReader : public AlembicIO
+				return 0.0;
+			}
+
+			ScalarPropertyReaderPtr scalarPropertyReader = propertyReader->getScalarProperty( name.string() );
```
Should we be stripping any prefix? Or just the "user:" prefix?

andrewkaufman commented on contrib/IECoreAlembic/src/IECoreAlembic/AlembicScene.cpp:
```diff
@@ -267,6 +312,601 @@ class AlembicScene::AlembicReader : public AlembicIO
+			const IXformSchema &schema = m_xform.getSchema();
+			auto userProperties = schema.getUserProperties();
+
+			const PropertyHeader *propertyHeader = userProperties.getPropertyHeader( name.string() );
```
Should we be stripping any prefix? Or just the "user:" prefix?

andrewkaufman commented on contrib/IECoreAlembic/src/IECoreAlembic/AlembicScene.cpp:
```diff
@@ -875,6 +1525,259 @@ class AlembicScene::AlembicWriter : public AlembicIO
+
+			OXformSchema &schema = m_xform.getSchema();
+
+			T prop( schema.getUserProperties(), name );
```
Should we be stripping any prefix? Or just the "user:" prefix?

johnhaddon commented on contrib/IECoreAlembic/src/IECoreAlembic/AlembicScene.cpp:
```diff
@@ -404,16 +428,40 @@ class AlembicScene::AlembicReader : public AlembicIO
+				return 0.0;
+			}
+
+			std::pair<index_t, chrono_t> floor = scalarPropertyReader->getFloorIndex( time );
```
Can we continue to use our `sampleInterval()` utility for this as we do for all the other interval queries? I'd be happier if we can, since we have some heuristics to avoid doing unnecessary interpolation when we're so near-as-dammit on a sample that it would just represent pointless overhead. It's also less code to maintain. From a quick squint at the headers it looks like we have the time sampling and number of samples available to us via scalarPropertyReader...

donboie replied:
Good point updated in b931144. @johnhaddon happy for me to squash & merge?

johnhaddon replied:
Yupyup - thanks Don!

donboie commented:
> Thanks for that - your parallelReadAll() stuff is coming in really handy! Just checking, the last line in your first table should be ".scc attributes" rather than ".abc attributes" right? So this is performing slightly better than SCC?

Yup abc is faster than scc for the attribute reads (I've updated the typo in the table) 

> In terms of round-tripping our own data through Alembic, I think just passing the names verbatim like we do in this PR is exactly what we need. But I'm not sure this is what people want if they're exporting custom attributes out of somewhere like Maya, and not using our naming convention. In that case there might be an argument for adding a "user:" prefix on read, so that the attribute is passed through to the renderer. But I don't know enough about common use patterns for Alembic to be sure. I suggest that for now we just move ahead as-is and see what arises. It might be that we're better off just documenting our convention, and suggesting that people use nodes in Gaffer to conform to it on read if they need to..

Perhaps we should have a section on the limitations of 3rd party format support in gaffer - usd, abc, vdb notes / known issues? What do you think @medubelko ?

johnhaddon commented:
> Perhaps we should have a section on the limitations of 3rd party format support in gaffer - usd, abc, vdb notes / known issues? What do you think @medubelko ?

Or a section on the capabilities of 3rd party support :) Sounds like a good idea to me, but before we can cover the specifics we're talking about here, I think we'll probably need sections on attributes in general, and how to use them to control shading in Arnold and elsewhere..

medubelko commented:
John has the right of it – first describe positively, not by negation. :)

I've added both to my list. Would attributes be something straightforward enough to research and tackle on my own, or should I outsource the primary writing portion, just to get it done faster?

johnhaddon commented:
> Would attributes be something straightforward enough to research and tackle on my own, or should I outsource the primary writing portion, just to get it done faster?

I think we need two things :

- The "Anatomy of a Scene" concepts document, where we describe attributes and their relation to the rest of the scene.
- A practical "using attributes to control shading" tutorial, probably focussed on Arnold
   - Storing texture paths in string attributes and referencing them in texture filename parameters
   - Passing color variation via attributes, perhaps with a mention of the application to instancing.

I feel like I've written dribs and drabs of the first one in various email threads (did it again today!), but never got it all in one place. Do you think you can collate that, or would it be better to outsource?
